### ISSUE TITLE
Excess properties warning with contextual type of `I | object`

### ISSUE THREAD

**Issue by andy-ms** [Microsoft/TypeScript#16235]
**TypeScript Version:**  nightly (2.4.0-dev.20170602)

**Code**

```ts
interface I { x: number }
declare function f(payload: I | object): void;
f({ foo: "bar" });
```

**Expected behavior:**

No error.

**Actual behavior:**

```ts
src/a.ts(3,5): error TS2345: Argument of type '{ foo: string; }' is not assignable to parameter of type 'object | I'.
  Object literal may only specify known properties, and 'foo' does not exist in type 'object | I'.
```

**Comment by sandersn** [comment]
checks for `isEmptyObject` should also check for `isPrimitiveObject`

**Comment by eliprand** [comment]
since this is a bug in `2.3.x` also, are there any plans to back port it?

**Comment by mhegazy** [comment]
no. this should be part of TS 2.4, an RC should be out early next week.

**Comment by eliprand** [comment]
ok. So, this was not considered a backward-compatibility-breaking bug? We have code that had been compiling just fine for months that broke on our build machine when it fetched `2.3.x`.
Just trying to understand how much I can trust TypeScript for production at this point...

**Comment by mhegazy** [comment]
Not sure we are talking about the same issue then. this error has been in 2.2, and 2.3. `object` was not available before 2.2, so maybe you are experiencing a different issue.

**Comment by eliprand** [comment]
@mhegazy you are correct. Looks like `mongodb` updated its definitions to use a custom `Default` type instead of `any` and use TypeScript `2.3`. The error manifested itself the same way as the `jsonwebtoken` issue I reported and I jumped to a conclusion. Thanks for looking into this.
